### DISCUSSION TITLE
Optimizing delivery of LL-HLS with NGINX

### DISCUSSION THREAD

**Discussion by mauricioabreu** [OvenMediaLabs/OvenMediaEngine#1689] [Discover]
My latest question here was about how to use NGINX along with OvenMediaEngine.

Today, I’m sharing my configuration here. It might help others bootstrap their own configurations.

```nginx
worker_processes  auto;

pid        /var/run/nginx.pid;

events {
    worker_connections  1024;
}

http {
    proxy_cache_path /opt/nginx_cache levels=1:2 keys_zone=cache_zone_one:10m;
    log_format main 'time="$time_iso8601" remote_addr="$remote_addr" req="$request" status="$status" '
                  'req_time="$request_time" bytes_sent="$body_bytes_sent" up_cache_status="$upstream_cache_status" '
                  'up_status="$upstream_status" up_addr"$upstream_addr" up_resp_time="$upstream_response_time" '
                  'ref="$http_referer" ua="$http_user_agent" scheme="$scheme"';
    access_log /dev/stdout main;

    upstream origin {
        server origin:3333;
        keepalive 64;
    }

    server {
        listen       80;
        server_name  localhost;
        proxy_ignore_headers "vary";
        proxy_hide_header "access-control-allow-origin";
        proxy_hide_header "access-control-allow-credentials";
        proxy_hide_header "access-control-allow-methods";
        proxy_hide_header "access-control-allow-headers";

        location ~ ^/.*.m3u8 {
            proxy_cache_key $request_uri;
            proxy_set_header Connection "";
            proxy_http_version 1.1;
            proxy_cache cache_zone_one;
            proxy_cache_lock on;
            add_header X-Cache-Status $upstream_cache_status;

            proxy_pass http://origin;
        }

        location ~ ^/.*.m4s {
            proxy_cache_key $request_uri;
            proxy_set_header Connection "";
            proxy_http_version 1.1;
            proxy_cache cache_zone_one;
            proxy_cache_lock on;
            add_header X-Cache-Status $upstream_cache_status;

            proxy_pass http://origin;
        }
    }
}
```

This configuration should have http2, but enabling locally is not straightforward - use [mkcert](https://github.com/FiloSottile/mkcert)

For those who are already familiar with LL-HLS and NGINX, what other settings or approaches would you recommend?

**Comment by naanlizard** [comment]
Are you certain you want to have your upstream connect to port 3333? That's typically the WebRTC port, not LLHLS

Can you explain your thinking with some comments inline? 

I'd also say your logging seems excessive, but perhaps that's what you want for debugging

**Comment by maruf-cs** [comment]
![image](https://github.com/user-attachments/assets/13c7f213-737c-4b5b-a6c3-b2d4f1736f87)

I noticed that the session (Yellow) appears to be unique for each user. By using "proxy_cache_key $request_uri;" the cache hit ratio decreases significantly. I'm trying to understand more about this session behavior—what exactly is it, and how is it managed for each user? If I skip the session in the proxy server, the stream doesn’t play smoothly. Could anyone provide some guidance on how to handle this efficiently?

**Reply by naanlizard** [reply]
I'm also curious about how the session key works.

I did find this guide to ignoring the session number but it sounds like you already tried that, how did you implement it?

How are you tracking your cache hit rate?

**Reply by getroot** [reply]
When there is an edge behind the origin, set LLHLS OriginMode to true. The session_key will be deleted, so it can be cached on the edge.

https://github.com/AirenSoft/OvenMediaEngine/blob/cdc0b62595170bf595c1dc9bb8d3e85b3f0c4957/misc/conf_examples/Origin.xml#L359

```
<LLHLS>
	<OriginMode>true</OriginMode>
```

**Reply by maruf-cs** [reply]
> I'm also curious about how the session key works.
> 
> I did find this guide to ignoring the session number but it sounds like you already tried that, how did you implement it?
> 
> How are you tracking your cache hit rate?

I tried manually parsing and ignoring the session parameter for the cache key, but it wasn’t the right approach. As for tracking the cache hit rate, I'm not actually tracking it directly at the moment. Instead, I'm monitoring the bandwidth IN and OUT ratio of the server as a rough indicator of how well the cache is performing.

**Reply by fatchan** [reply]
> When there is an edge behind the origin, set LLHLS OriginMode to true. The session_key will be deleted, so it can be cached on the edge.
> 
> https://github.com/AirenSoft/OvenMediaEngine/blob/cdc0b62595170bf595c1dc9bb8d3e85b3f0c4957/misc/conf_examples/Origin.xml#L359
> 
> ```
> <LLHLS>
> 	<OriginMode>true</OriginMode>
> ```

This should really be mentioned in the docs (https://docs.ovenmediaengine.com/configuration). Unless I missed it I couldn't find it from a quick search. Solved my problem with being able to cache video chunks. Much appreciated.

**Reply by getroot** [reply]
@fatchan I would appreciate it if you could PR to add it to the docs.

**Reply by fatchan** [reply]
I am happy to do this, although I'm not 100% sure on what OriginMode does in detail. From what I can tell, it removes the session IDs from the query string in the requests, so all viewers are requesting the same URLs and I am able to cache the video chunks.
Do the session IDs also have something to do with ABR? In my use case I'm not using ABR so this is fine, but if it's not compatible with OriginMode I would need to mention that in the docs.

**Comment by naanlizard** [comment]
Another topic here

Recently, we tried to deploy a change that would serve thumbs from OME, rather than our bespoke thumbnailing process. We use nginx as a reverse proxy, and we had ome set as an upstream like so:

```
upstream ome {
        server internalip:800;
        keepalive 32; # Note keepalive
    }
```

Some other relevant config:

`proxy_set_header Connection $connection_upgrade;` and `proxy_http_version 1.1;` in our proxy.conf (this is largely the default docker-swag config)

```
    location /live {

        include /config/nginx/proxy.conf;
        include /config/nginx/securityheaders.conf;

        proxy_cache ome_cache;
        proxy_cache_valid 200 302 1h;
        proxy_cache_use_stale error timeout http_500 http_502 http_503 http_504;
        proxy_no_cache $arg_nocache;
        add_header X-Cache-Status $upstream_cache_status;
        proxy_cache_key $scheme$proxy_host$request_uri$http_origin;
        proxy_cache_lock on;
        proxy_cache_lock_timeout 2s;

        # Don't verify OME's ssl cert, it wouldn't pass as the hostname is incorrect
        proxy_ssl_verify off;

        proxy_pass https://ome/live;
```

Note, at no point do we add the following line, which I believe is a mistake, since the [nginx docs say for keepalive connections you want to clear the Connection header](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive)

`        proxy_set_header Connection "";`

The keepalive is intended to improve performance by reusing connections rather than making/closing connections every time

Before the attempted deploy, the only thing that uses this upstream was the playback endpoint, which worked great. 

We deployed an update that added another location that proxy passed requests for images to OME, for the thumbnails

This eventually caused issues for stream playback and resulted in 404s for stream playlists (m3u8 files), which were a huge pain in the ass to track down. My guess is that either OME or nginx don't like requesting different content types over the same connection, or something along those lines, or perhaps a header issue (one is served from a location like hostname.tld/ome-screenshots, the other from a subdomain). 

To fix this, we've made a separate upstream definition, added the line to clear the connection header to the playback location (but not the image upstream, which I suppose might cause issues, now that I'm thinking about it?) and more recently I've disabled keepalive entirely for the images upstream (since I assume the performance gains there are minimal)

Note that with 32 workers, and a keepalive of 32, that means we'll have up to 32ˆ2 connections to OME, which I'm not sure is ideal.

How have the rest of you sorted nginx proxying? I'm sure we can improve on this, but I'm not fantastic with nginx. Do we even care about keepalives between nginx and OME on the same physical server (and different docker containers?) 

Related, how does OME determine a client is a different client or not? It would be nice to have accurate client counts, but filtered through nginx, I'm not sure we could

For reference, here's our current upstream set

```
    upstream ome-images {
	    server internalip:800;
    }

    upstream ome-playback {
	    server internalip:800;
	    keepalive 32;
    }
```

**Reply by naanlizard** [reply]
Following up with this, it seems to have continued even after this config change, and ultimately we had to disable the image endpoint we had created - after that, HLS playback returned to normal. I'll try keepalives again in a while

For the moment, I'm using a simple bash script to copy the thumbnails to disk and then serve them like usual with nginx

**Reply by fatchan** [reply]
I hope this isn't too late of a reply but,

I noticed in your separate upstreams for images and playback, you use the same port. From my understanding, OME has a separate port for these. I haven't tried using the same port, but setting a different port and using separate upstream addresses works flawlessly for me 👍🏻 

<details>
<summary>An example using the same domain, with a rewrite so the path for thumbs starts with /thumb:</summary>

```
upstream ovenmedia_hls {
  server 127.0.0.1:3338;
  keepalive 1000;
}

upstream ovenmedia_thumb {
  server 127.0.0.1:3340;
  keepalive 100;
}

# other stuff ...

server {
  listen 443 ssl http2;

  # other stuff ...

  location /thumb/ {
    rewrite ^/thumb/(.*)$ /$1 break;
    proxy_pass http://ovenmedia_thumb;
    proxy_set_header Host $host;
    proxy_http_version 1.1;
    proxy_set_header Connection "";
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  }

  location / {
    proxy_pass http://ovenmedia_hls;
    proxy_buffering off;
    proxy_set_header Host $host;
    proxy_http_version 1.1;
    proxy_set_header Connection "";
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  }

  # other stuff ...

}
```

</details>
